packages/core/react/src/WalletProvider.tsx

```diff
@@ -15,14 +22,69 @@ export interface WalletProviderProps {
     localStorageKey?: string;
 }
 
+let _userAgent: string | null;
+function getUserAgent() {
+    if (_userAgent === undefined) {
+        _userAgent = globalThis.navigator?.userAgent ?? null;
+    }
+    return _userAgent;
+}
+
+function getIsMobile(adapters: Adapter[]) {
+    const userAgentString = getUserAgent();
+    return getEnvironment({ adapters, userAgentString }) === Environment.MOBILE_WEB;
+}
+
+function getUriForAppIdentity() {
+    const location = globalThis.location;
+    if (location == null) {
+        return;
+    }
+    return `${location.protocol}//${location.host}`;
+}
+
 export function WalletProvider({
     autoConnect,
     localStorageKey = 'walletName',
     wallets: adapters,
     ...props
 }: WalletProviderProps) {
-    const [walletName, setWalletName] = useLocalStorage<WalletName | null>(localStorageKey, null);
-    const adapter = useMemo(() => adapters.find((a) => a.name === walletName) ?? null, [adapters, walletName]);
+    const mobileWalletAdapter = useMemo(() => {
+        if (!getIsMobile(adapters)) {
+            return null;
+        }
+        const existingMobileWalletAdapter = adapters.find(
+            (adapter) => adapter.name === SolanaMobileWalletAdapterWalletName
+        );
+        if (existingMobileWalletAdapter) {
+            return existingMobileWalletAdapter;
+        }
+        return new SolanaMobileWalletAdapter({
+            addressSelector: createDefaultAddressSelector(),
+            appIdentity: {
+                uri: getUriForAppIdentity(),
+            },
+            authorizationResultCache: createDefaultAuthorizationResultCache(),
+            cluster: 'mainnet-beta',
+        });
+    }, [adapters]);
+    const adaptersWithDefaultsInjected = useMemo(() => {
+        if (!getIsMobile(adapters)) {
+            return adapters;
+        }
+        if (adapters.indexOf(mobileWalletAdapter as Adapter) !== -1) {
+            return adapters;
+        }
+        return [mobileWalletAdapter, ...adapters] as Adapter[];
+    }, [adapters, mobileWalletAdapter]);
+    const [walletName, setWalletName] = useLocalStorage<WalletName | null>(
+        localStorageKey,
+        getIsMobile(adapters) ? SolanaMobileWalletAdapterWalletName : null
```

Comment:
If on mobile, this makes sure that the mobile wallet name is selected by default if we don't already have a cached wallet name in `localStorage`.